pandas/_libs/tslibs/util.pxd

```diff
@@ -227,3 +236,44 @@ cdef inline bint is_nan(object val):
     is_nan : bool
     """
     return (is_float_object(val) or is_complex_object(val)) and val != val
+
+
+cdef inline bint get_string_data(object s, const char **buf,
+                                 Py_ssize_t *length):
+    """
+    Extract internal char * buffer of unicode or bytes object `s` to `buf` with
+    getting length of this internal buffer saved in `length`.
+    Returns `False` if it failed to extract such buffer for whatever reason,
+    otherwise returns `True`.
+
+    Notes
+    -----
+    Python object owns memory, `buf` should not be freed.
+    `length` can be NULL if getting buffer length is not needed.
+    This function should only raise exceptions in out-of-memory cases.
+
+    Parameters
+    ----------
+    s      : object
+    buf    : const char**
+    length : Py_ssize_t*
+
+    Returns
+    -------
+    bint
+    """
+    if PyUnicode_Check(s):
+        buf[0] = PyUnicode_AsUTF8AndSize(s, length)
+        return buf[0] != NULL
+    if PyBytes_Check(s):
+        return PyBytes_AsStringAndSize(s, <char**>buf, length) == 0
+    return False
+
+cdef inline void get_string_data_checked(object s, const char **buf,
+                                         Py_ssize_t *length):
+    """
+    This is a wrapper for get_string_data() that raises TypeError
+    when supplied with neither unicode nor bytes object
+    """
+    if not get_string_data(s, buf, length):
+        PyErr_BadArgument()
```

Comment:
is this just the C version of `raise TypeError`?

Reply:
Yes, but also with a correct message (this calls exactly the same function as would `PyUnicode_AsUTF8AndSize` call, so the exception is identical to what original function would set).

Reply:
Yes.

Reply:
We fixed the code, so this call is no longer needed